torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py

```diff
@@ -65,6 +65,51 @@ def decompress(fut):
     return fut.then(decompress)
 
 
+class OptimizerHookState(object):
+    """
+    Holds state for running optimizer in-line after DDP communication hook.
+    Currently contains optimizer class as well as optimizer args.
+    """
+    __slots__ = ['functional_optim_cls', 'functional_optim_args', 'functional_optimizer']
+
+    def __init__(self, functional_optim_cls, *functional_optim_args):
+        # TODO: Support kwargs
+        self.functional_optim_cls = functional_optim_cls
+        self.functional_optim_args = functional_optim_args
+        self.functional_optimizer = self.functional_optim_cls(
+            [],
+            *self.functional_optim_args,
+            allow_empty_param_list=True
+        )
+        if not hasattr(self.functional_optimizer, 'step_param'):
+            raise ValueError(
+                f"Class {self.functional_optim_cls} must implement method step_param."
+            )
+
+def hook_then_optimizer(
+    hook: Callable[[Any, dist.GradBucket], torch.futures.Future], optimizer_state: OptimizerHookState
+) -> Callable[[Any, dist.GradBucket], torch.futures.Future]:
+    """Runs optimizer in a functional fashion after DDP communication hook."""
+
+    def hook_then_optimizer_wrapper(
+        hook_state, bucket: dist.GradBucket
+    ) -> torch.futures.Future:
+        # Run original hook
+        fut = hook(hook_state, bucket)
+
+        def optimizer_step(fut):
+            gradient_tensors = bucket.get_per_parameter_tensors()
+            model_params = bucket.get_model_params_for_bucket()
+            for grad_tensor, model_param in zip(gradient_tensors, model_params):
+                optimizer_state.functional_optimizer.step_param(
+                    model_param,
+                    grad_tensor,
+                )
+            return [bucket.get_tensor()]
```

Comment:
FYI: This should be a single tensor after #62074

Reply:
Sounds good, I might still run a few more experiments/polishing, I'll rebase if you land that PR first